test: wallet rescan with reorged parent + IsFromMe child in mempool

glozow opened the pull request:
Originally motivated by #29019, which reverts back to having `requestMempoolTransactions` emit `transactionAddedToMempool` in `mapTx` default order instead of `GetSortedDepthAndScore` order.

It's important that these notifications happen in topological order, otherwise the wallet rescan may miss transactions that belong to it. Notably, checking whether a transaction `IsFromMe` requires knowing its inputs, which may be from a mempool parent.

When using `mapTx` order, a parent may come later than its child if it was added from a block disconnected in a reorg.

This PR adds a test for this case.

glozow committed:
[test] rescan legacy wallet with reorged parent + IsFromMe child in mempool

Test that wallet rescans process transactions topologically, even if a
parent's entry into the mempool is later than that of its child.
This behavior is important because IsFromMe requires the ability to look
up a transaction's inputs.

achow101 commented:
Huh, looks like this behavior has been broken since it was introduced in #15652

achow101 commented:
ACK 1a52ca7619eeb1baafa5a32b364381126862b29d

stickies-v commented on test/functional/wallet_import_rescan.py:
```diff
@@ -79,7 +82,7 @@ def do_import(self, timestamp):
             assert_equal(response, [{"success": True}])
 
-    def check(self, txid=None, amount=None, confirmation_height=None):
+    def check(self, txid=None, amount=None, confirmation_height=None, amount_is_balance=True):
```
It doesn't seem like we're actually testing the expected balance at any point, rather the amount received, but I may be missing nuance. Would this diff make sense? I don't like adding an extra parameter to an already long function for something so specific, if we can avoid it.

<details>
<summary>git diff on 1a52ca7619</summary>

```diff
diff --git a/test/functional/wallet_import_rescan.py b/test/functional/wallet_import_rescan.py
index 1abc9c0c79..fa9f43f66c 100755
--- a/test/functional/wallet_import_rescan.py
+++ b/test/functional/wallet_import_rescan.py
@@ -82,7 +82,7 @@ class Variant(collections.namedtuple("Variant", "call data address_type rescan p
             )
             assert_equal(response, [{"success": True}])
 
-    def check(self, txid=None, amount=None, confirmation_height=None, amount_is_balance=True):
+    def check(self, txid=None, amount=None, confirmation_height=None):
         """Verify that listtransactions/listreceivedbyaddress return expected values."""
 
         txs = self.node.listtransactions(label=self.label, count=10000, include_watchonly=True)
@@ -112,8 +112,7 @@ class Variant(collections.namedtuple("Variant", "call data address_type rescan p
 
             address, = [ad for ad in addresses if txid in ad["txids"]]
             assert_equal(address["address"], self.address["address"])
-            if amount_is_balance:
-                assert_equal(address["amount"], self.expected_balance)
+            assert_equal(address["amount"], self.amount_received)
             assert_equal(address["confirmations"], confirmations)
             # Verify the transaction is correctly marked watchonly depending on
             # whether the transaction pays to an imported public key or
@@ -227,11 +226,11 @@ class ImportRescanTest(BitcoinTestFramework):
             variant.node = self.nodes[2 + IMPORT_NODES.index(ImportNode(variant.prune, expect_rescan))]
             variant.do_import(variant.timestamp)
             if expect_rescan:
-                variant.expected_balance = variant.initial_amount
+                variant.amount_received = variant.initial_amount
                 variant.expected_txs = 1
                 variant.check(variant.initial_txid, variant.initial_amount, variant.confirmation_height)
             else:
-                variant.expected_balance = 0
+                variant.amount_received = 0
                 variant.expected_txs = 0
                 variant.check()
 
@@ -251,7 +250,7 @@ class ImportRescanTest(BitcoinTestFramework):
         # Check the latest results from getbalance and listtransactions.
         for variant in IMPORT_VARIANTS:
             self.log.info('Run check for variant {}'.format(variant))
-            variant.expected_balance += variant.sent_amount
+            variant.amount_received += variant.sent_amount
             variant.expected_txs += 1
             variant.check(variant.sent_txid, variant.sent_amount, variant.confirmation_height)
 
@@ -297,7 +296,6 @@ class ImportRescanTest(BitcoinTestFramework):
                 subtract_fee_from_outputs=[0]
             )
             variant.child_txid = child["txid"]
-            variant.expected_balance = 0
             self.nodes[0].sendrawtransaction(child["hex"])
 
         # Mempools should contain the child transactions for each variant.
@@ -321,10 +319,11 @@ class ImportRescanTest(BitcoinTestFramework):
                 # Ensure both transactions were rescanned
                 assert_equal(variant.node.gettransaction(variant.initial_txid)['confirmations'], 0)
                 assert_equal(variant.node.gettransaction(variant.child_txid)['confirmations'], 0)
+                variant.amount_received = variant.initial_amount
                 variant.expected_txs = 1
-                variant.check(variant.initial_txid, variant.initial_amount, 0, False)
+                variant.check(variant.initial_txid, variant.initial_amount, 0)
             else:
-                variant.expected_balance = 0
+                variant.amount_received = 0
                 variant.expected_txs = 0
                 variant.check()
 

```
</details>

glozow replied:
Thanks, taken.

stickies-v reviewed (commented):
Approach ACK, nice one finding this test case! 1a52ca7619eeb1baafa5a32b364381126862b29d looks good to me but not super familiar with this part of the code. Left some suggestions about maintainability and can be ignored.

I verified that the tests fail on #29019 as well as on 453b4813ebc74859864803e9972b58e4be76a4d6~1, as is necessary.

glozow commented on test/functional/wallet_import_rescan.py:
```diff
@@ -267,14 +271,45 @@ def run_test(self):
+        # Mine a block so these parents are confirmed
+        assert_equal(len(self.nodes[0].getrawmempool()), len(mempool_variants))
+        self.sync_mempools()
+        block_to_disconnect = self.generate(self.nodes[0], 1)
```
I agree, taken

glozow commented on test/functional/wallet_import_rescan.py:
```diff
@@ -267,14 +271,45 @@ def run_test(self):
+        # belonging to this wallet when the inputs are known. That means the rescan must process the
+        # parent before the child.
+        for variant in mempool_variants:
+            inputs_to_spend = [txin for txin in self.nodes[1].listunspent() if txin["txid"] == variant.initial_txid]
```
taken

glozow commented on test/functional/wallet_import_rescan.py:
```diff
@@ -283,9 +318,11 @@ def run_test(self):
             if expect_rescan:
-                variant.expected_balance = variant.initial_amount
+                # Ensure both transactions were rescanned
+                assert_equal(variant.node.gettransaction(variant.initial_txid)['confirmations'], 0)
+                assert_equal(variant.node.gettransaction(variant.child_txid)['confirmations'], 0)
```
Added a similar comment.

furszy reviewed (commented):
Pre-review note: have you considered that `wallet_import_rescan.py` is a legacy wallet only test?
If it's placed here due to a lack of alternatives, I think this could be moved to a separate `wallet_mempool.py` file, where we could continue adding more cases related to the wallet-mempool interaction (or.. we could upgrade this file to run on a descriptors wallet).

glozow commented:
> Pre-review note: have you considered that wallet_import_rescan.py is a legacy wallet only test?

Ah good point. I added a commit to run this test with descriptors as I couldn't find a reason why we don't do this. Lmk if this seems sufficient / ok to do?

glozow commented:
I'm working on adding a descriptors version of this test so we can have both.

glozow committed:
[test] import descriptor wallet with reorged parent + IsFromMe child in mempool

Test that wallet rescans process transactions topologically, even if a
parent's entry into the mempool is later than that of its child.
This behavior is important because IsFromMe requires the ability to look
up a transaction's inputs.

Co-authored-by: furszy <matiasfurszyfer@protonmail.com>

glozow commented:
I've added another commit which has pretty much the same test but using descriptor wallets i.e. a rescan through `importdescriptors`. Just like the other test, you can reproduce the bug by cherry-picking from #29019. You should get `JSONRPCEXCEPTION: Invalid or non-wallet transaction id (-5)` for the child tx.

stickies-v commented on test/functional/wallet_rescan_unconfirmed.py:
```diff
@@ -0,0 +1,97 @@
+
+        self.log.info("Create wallets and mine initial chain")
+        tester_wallet = MiniWallet(tester_node)
+        self.generate(tester_wallet, 110)
```
Is there a reason it's 110 and not `COINBASE_MATURITY`?

glozow replied:
changed

stickies-v commented on test/functional/wallet_rescan_unconfirmed.py:
```diff
@@ -0,0 +1,97 @@
+        # Check that parent address is correctly determined as ismine
+        test_address(w1, parent_address, solvable=True, ismine=True)
+        # This would raise a JSONRPCError if the transactions were not identified as belonging to the wallet.
+        assert_equal(w1.gettransaction(tx_parent_to_reorg["txid"])["confirmations"], 0)
+        assert_equal(w1.gettransaction(tx_child_unconfirmed_sweep["txid"])["confirmations"], 0)
```
nit: I still think it's unnecessarily confusing for (current and future) reviewers to figure out why confirmations needs to equal 0 when afaiu this is completely unrelated to the test. I would either remove the value comparison or specifically state in the docs that we don't care about the `confirmations` value, just that the response is a JSON object with a field we recognize?

glozow replied:
We are testing that they are marked as in-mempool wallet transactions.

stickies-v replied:
Makes sense, thank you! I failed to appreciate that we're calling the wallet `gettransaction` RPC, and not the not `getrawtransaction` RPC.

glozow replied:
Ah yeah, it's a wallet-specific RPC. I've added a check that the "confirmations" value is 1 on wallet 0 a few lines up, prior to the reorg. Hopefully that also helps make it clear why we're testing this way.

stickies-v reviewed (commented):
Hooray for adding descriptor wallet support. Is there a reason both legacy and descriptor wallet tests aren't in the new `wallet_rescan_unconfirmed` test, though? That seems like a more logical grouping with less duplication? If there are technical objections, I think it would be good to add some extra documentation to both tests stating that the other wallet type is tested in file xxx.py, making it easier to clean up these tests e.g. when we drop legacy wallets or make other significant changes?

glozow commented:
> Is there a reason both legacy and descriptor wallet tests aren't in the new wallet_rescan_unconfirmed test, though? That seems like a more logical grouping with less duplication?

They use different RPCs and different codepaths, so I don't really see why they need to be in the same file or tied together. I think it's more readable for them to be separate (having most the test as `if is_sqlite_compiled() elif is_bdb_compiled()` is pretty annoying). I think it's much easier to delete legacy without touching the descriptor one if they're in different files.

glozow commented on test/functional/wallet_rescan_unconfirmed.py:
```diff
@@ -0,0 +1,97 @@
+        self.skip_if_no_sqlite()
+
+    def run_test(self):
+        # Node with which we will use to send funds to the wallet and mine blocks
```
taken

Eunovo commented:
Tested ACK [d9df438](https://github.com/bitcoin/bitcoin/pull/29179/commits/d9df438c6e581dae0c818a4c2f5fe95627ae26bc)

Ran the tests locally and got `test_framework.authproxy.JSONRPCException: Invalid or non-wallet transaction id (-5)` for both `wallet_import_rescan` and `wallet_rescan_unconfirmed` when random order is used in `CTxMemPool::entryAll()`.

stickies-v commented on test/functional/wallet_rescan_unconfirmed.py:
```diff
@@ -0,0 +1,98 @@
+        self.log.info("Import descriptor wallet on another node")
+        descriptors_to_import = []
+        for item in w0.listdescriptors()["descriptors"]:
+            # descriptors_to_import.append(item)
```
probably should be removed?

glozow replied:
Oops! Removed.

stickies-v commented on test/functional/wallet_import_rescan.py:
```diff
@@ -267,14 +270,44 @@ def run_test(self):
+        # For each variant, create an unconfirmed child transaction from initial_txid, sending all
+        # the funds to an unspendable address. Importantly, no change output is created so the
+        # transaction can't be recognized using its outputs. The wallet rescan needs to know the
+        # inputs of the transaction to detect it, so the parent must be processed before the child.
```
What do you think about adding one line to these docs in each file to reference the other file? Might be useful to e.g. keep tests in sync if problems are found, or when people wonder why this is (at first sight) only tested for legacy or descriptor wallets.

```suggestion
        # inputs of the transaction to detect it, so the parent must be processed before the child.
        # This behaviour is tested for descriptor wallets in wallet_rescan_unconfirmed.py.
```

glozow replied:
Added this comment in the legacy test as I agree it'll be convenient to know what coverage we are/aren't losing when deleting a legacy test.

stickies-v reviewed (approved):
ACK d9df438c6e581dae0c818a4c2f5fe95627ae26bc

> I think it's much easier to delete legacy without touching the descriptor one if they're in different files.

Thanks, you're right. There's less overlap than I thought there would be.

stickies-v reviewed (approved):
re-ACK c39e82926a3740642f8f1bb72c6b10b67a6dc724

glozow commented:
CI failure is unrelated, #29234

furszy commented on test/functional/wallet_rescan_unconfirmed.py:
```diff
@@ -0,0 +1,98 @@
+
+
+        self.log.info("Create wallets and mine initial chain")
+        tester_wallet = MiniWallet(tester_node)
+        self.generate(tester_wallet, COINBASE_MATURITY)
```
Unless you specify a clean chain (which this test does not), the test will have an existing matured. Which means, you could modify this for 

```suggestion
        tester_wallet = MiniWallet(tester_node)
        self.generate(tester_wallet, 1)
```

glozow replied:
Keeping as is because I prefer the peace of mind of knowing `tester_wallet` has a mature coinbase.

furszy replied:
> Keeping as is because I prefer the peace of mind of knowing `tester_wallet` has a mature coinbase.

You shouldn't be concerned about that. Thats how the test framework behaves when `setup_clean_chain` is not set. See [code](https://github.com/bitcoin/bitcoin/blob/3ba8de1b704d590fa4e1975620bd21d830d11666/test/functional/test_framework/test_framework.py#L381). If the wallet wouldn't be having mature coins, the test would be always failing due a lack of funds.

glozow replied:
Deleted this line altogether, there's no reason to generate 1 either.

furszy reviewed (commented):
Code reviewed, looking good. Left few comments, mostly testing time improvements. Nothing blocking.

As the scenario being covered does not require verifying p2p mempool relay, you could speed up the new `wallet_rescan_unconfirmed.py` test by ~2.5x (tested locally, ~3 secs versus ~8 secs) by using a single node instead of creating three of them. See commit f8273512 (feel free to pull it).

glozow commented on test/functional/wallet_rescan_unconfirmed.py:
```diff
@@ -0,0 +1,98 @@
+
+
+class WalletRescanUnconfirmed(BitcoinTestFramework):
+    def add_options(self, parser):
+        self.add_wallet_options(parser, legacy=False)
+
+    def set_test_params(self):
+        self.num_nodes = 3
+        # Immediate tx relay
+        self.extra_args = [['-whitelist=noban@127.0.0.1']] * self.num_nodes
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+        self.skip_if_no_sqlite()
```
Seems pretty common to specify both. They are technically different; one says no legacy and the other says need descriptors. I know that's effectively the same thing and we don't have 3 types, but I think it's harmless to keep this.

glozow commented on test/functional/wallet_rescan_unconfirmed.py:
```diff
@@ -0,0 +1,98 @@
+        assert_equal(len(w0_utxos), 1)
+        assert_equal(w0_utxos[0]["txid"], tx_parent_to_reorg["txid"])
+        tx_child_unconfirmed_sweep = w0.sendall([ADDRESS_BCRT1_UNSPENDABLE])
+        self.sync_all()
```
done

glozow commented on test/functional/wallet_rescan_unconfirmed.py:
```diff
@@ -0,0 +1,98 @@
+        for item in w0.listdescriptors()["descriptors"]:
+            # use timestamp 0 to tell wallet to rescan entire chain
+            if not item["internal"]:
+                descriptors_to_import.append({"desc": item["desc"], "timestamp": 0, "label": "w0 import"})
```
nice, taken

glozow commented on test/functional/wallet_rescan_unconfirmed.py:
```diff
@@ -0,0 +1,98 @@
+            # use timestamp 0 to tell wallet to rescan entire chain
+            if not item["internal"]:
+                descriptors_to_import.append({"desc": item["desc"], "timestamp": 0, "label": "w0 import"})
+        import_wallet_node.createwallet(wallet_name="w1", disable_private_keys=True, blank=True, descriptors=True)
```
dropped the args

glozow commented on test/functional/wallet_rescan_unconfirmed.py:
```diff
@@ -0,0 +1,98 @@
+        tester_wallet = MiniWallet(tester_node)
+        self.generate(tester_wallet, COINBASE_MATURITY)
+
+        orig_wallet_node.createwallet(wallet_name='w0', disable_private_keys=False, descriptors=True)
```
dropped the extra args

stickies-v reviewed (approved):
re-ACK 9db0d42509e0e223b45238fd8087fc98ef32c091

glozow commented:
> See commit https://github.com/bitcoin/bitcoin/commit/f8273512a50d9ed03bc48803634fb8780b9023c7 (feel free to pull it).

Sorry I hadn't seen this. Thanks, taken with a couple of edits.

furszy commented on test/functional/wallet_rescan_unconfirmed.py:
```diff
@@ -0,0 +1,83 @@
+        # inputs of the transaction to detect it, so the parent must be processed before the child.
+        w0_utxos = w0.listunspent()
+
+        self.log.info("Create a child tx and wait for it to propagate to all mempools")
```
nit:
leftover
```suggestion
        self.log.info("Create a child tx and wait until all wallets are notified")
```

furszy commented on test/functional/wallet_rescan_unconfirmed.py:
```diff
@@ -0,0 +1,83 @@
+        node.invalidateblock(block_to_reorg)
+        assert tx_parent_to_reorg["txid"] in node.getrawmempool()
+
+        self.log.info("Import descriptor wallet on another node")
```
nit:
leftover; should be "import descriptor on another wallet"

furszy reviewed (commented):
Code review ACK df30247705, nits can be disregarded.

achow101 commented:
ACK df30247705940c50c5eaafd74e2abbeb8b0cec07

stickies-v reviewed (commented):
Post-merge re-ACK df30247705940c50c5eaafd74e2abbeb8b0cec07, I was still investigating the non-deterministic behaviour in `wallet_import_rescan` but I've now found the issue and it doesn't seem problematic enough to revert or immediately fix.

When rebasing these tests on top of e.g. https://github.com/bitcoin/bitcoin/pull/29019 as well as on https://github.com/bitcoin/bitcoin/commit/453b4813ebc74859864803e9972b58e4be76a4d6~1, the [mempool variant tests](https://github.com/bitcoin/bitcoin/pull/29179/files#diff-94e0490035bf63bb005387bd8e89e721762b1424ae9f53248a1e3fbef6f763ddR322-R323) should fail for every variant where `expect_rescan == True` because in each case, _only_ the parent is re-imported in the mempool (after re-org). However, when running the test multiple times one can observe this is not the case, and it sometimes takes 2 or even more rescan variants before the test fails.

<details>
<summary>log outputs</summary>

Expected failure logs:
```
...
2024-01-18T13:20:15.028000Z TestFramework (INFO): Test that the mempool is rescanned as well if the rescan parameter is set to true
2024-01-18T13:20:30.271000Z TestFramework (INFO): Run import for mempool variant Variant(call=<Call.single: 1>, data=<Data.address: 1>, address_type=<AddressType.bech32: 'bech32'>, rescan=<Rescan.no: 1>, prune=False)
2024-01-18T13:20:30.317000Z TestFramework (INFO): Run import for mempool variant Variant(call=<Call.single: 1>, data=<Data.address: 1>, address_type=<AddressType.bech32: 'bech32'>, rescan=<Rescan.yes: 2>, prune=False)
2024-01-18T13:20:30.388000Z TestFramework (ERROR): JSONRPC error
...
```

Actual failure logs, sometimes:
```
...
2024-01-18T13:25:50.799000Z TestFramework (INFO): Test that the mempool is rescanned as well if the rescan parameter is set to true
2024-01-18T13:26:06.108000Z TestFramework (INFO): Run import for mempool variant Variant(call=<Call.single: 1>, data=<Data.address: 1>, address_type=<AddressType.bech32: 'bech32'>, rescan=<Rescan.no: 1>, prune=False)
2024-01-18T13:26:06.164000Z TestFramework (INFO): Run import for mempool variant Variant(call=<Call.single: 1>, data=<Data.address: 1>, address_type=<AddressType.bech32: 'bech32'>, rescan=<Rescan.yes: 2>, prune=False)
2024-01-18T13:26:06.235000Z TestFramework (INFO): Run import for mempool variant Variant(call=<Call.single: 1>, data=<Data.address: 1>, address_type=<AddressType.p2sh_segwit: 'p2sh-segwit'>, rescan=<Rescan.no: 1>, prune=False)
2024-01-18T13:26:06.268000Z TestFramework (INFO): Run import for mempool variant Variant(call=<Call.single: 1>, data=<Data.address: 1>, address_type=<AddressType.p2sh_segwit: 'p2sh-segwit'>, rescan=<Rescan.yes: 2>, prune=False)
2024-01-18T13:26:06.341000Z TestFramework (INFO): Run import for mempool variant Variant(call=<Call.single: 1>, data=<Data.address: 1>, address_type=<AddressType.legacy: 'legacy'>, rescan=<Rescan.no: 1>, prune=False)
2024-01-18T13:26:06.374000Z TestFramework (INFO): Run import for mempool variant Variant(call=<Call.single: 1>, data=<Data.address: 1>, address_type=<AddressType.legacy: 'legacy'>, rescan=<Rescan.yes: 2>, prune=False)
2024-01-18T13:26:06.449000Z TestFramework (INFO): Run import for mempool variant Variant(call=<Call.single: 1>, data=<Data.pub: 2>, address_type=<AddressType.bech32: 'bech32'>, rescan=<Rescan.no: 1>, prune=False)
2024-01-18T13:26:06.494000Z TestFramework (INFO): Run import for mempool variant Variant(call=<Call.single: 1>, data=<Data.pub: 2>, address_type=<AddressType.bech32: 'bech32'>, rescan=<Rescan.yes: 2>, prune=False)
2024-01-18T13:26:06.578000Z TestFramework (ERROR): JSONRPC error
...
```
</details>

This happens because  a chain of transactions [can be formed](https://github.com/bitcoin/bitcoin/pull/29179/files#diff-94e0490035bf63bb005387bd8e89e721762b1424ae9f53248a1e3fbef6f763ddR274), where `variant_a.child_tx == variant_b.initial_txid`, causing `variant_a`'s child tx to be re-imported into the mempool when `variant b`'s initial tx is re-orged.

I don't think this needs to be fixed or reverted, because if such a chain is formed, the last variant in the chain is still guaranteed to fail. The descriptor test is unaffected because we create only a single parent/child combo.

stickies-v commented:
Reviewers might be interested in https://github.com/bitcoin/bitcoin/pull/29283, fixing a rare intermittency issue
